sound/soc/sof/pcm.c

```diff
@@ -395,6 +395,9 @@ static int sof_pcm_open(struct snd_pcm_substream *substream)
 	if (rtd->dai_link->no_pcm)
 		return 0;
 
+	if (!try_module_get(sdev->dev->driver->owner))
+		return -ENODEV;
+
 	spcm = snd_sof_find_spcm_dai(sdev, rtd);
 	if (!spcm)
 		return -EINVAL;
```

Comment:
so right there the code is wrong. If you return any error during the open, the close will not be called and you will not release the module.
You need to fix the error flow.

Reply:
I might be wrong, but wouldn't it be better to use pci_dev_get() / get_device()? In fact such explicit refcount increments are rarely needed, I think, usually with device hierarchies?

Reply:
I must admit I don't fully understand the links between module refcount and device refcount. @ranj063 made a mention that the two are related but I don't have more information or knowledge on this.
The main issue is that we don't have a real hierarchy here. The PCM open happens at the card level, which is a platform device created by the PCI driver. It's a bit ugly.

Reply:
@plbossart @zhuyingjiang I agree with @lyakh I think it is generally a bad idea to increment the refcount for the pci device from itself. The reason we have this problem is because the pcm component driver's owner module is the machine driver and only its refcount is increased during pcm open. 

I think a better way to tackle this is to adjust our sequence in snd_sof_device_remove(). Maybe we should not unregister the component if the pcm is open? Just thinking aloud

Reply:
> so right there the code is wrong. If you return any error during the open, the close will not be called and you will not release the module.
> You need to fix the error flow.

@plbossart Yes, thanks. Will update.

Reply:
> @plbossart @zhuyingjiang I agree with @lyakh I think it is generally a bad idea to increment the refcount for the pci device from itself. The reason we have this problem is because the pcm component driver's owner module is the machine driver and only its refcount is increased during pcm open.
> 
> I think a better way to tackle this is to adjust our sequence in snd_sof_device_remove(). Maybe we should not unregister the component if the pcm is open? Just thinking aloud

explicit refcount increment of the module is used by ALSA pcm(see pcm_native.c/pcm_oss.c).
Since SOF defined an spcm itself, we need make this aligned.

We start the module unload from sof_pci_dev, as @keyonjie said "once sof_pci_dev is removed, all other SOF related module can be removed", so when playing we need protect the sof_pci_dev not to be removed by increase the refcount.

The lsmod diff of HDA attached as reference.

Reply:
@zhuyingjiang what you are referring to is the core incrementing the card module's ref count.
But what we're trying to do here is incrementing the sof_pci_dev's refcount within the SOF driver.
I'm not sure about how this will be received upstream.

Reply:
@ranj063 the module refcount and device refcount seems are not the same:
dev->kobj->kref->refcount
device_driver->module->refcnt
@lgirdwood how will this case handle in ASoC drivers, any suggestions with this?

Reply:
Interesting case, I agree with @zhuyingjiang that this needs to be handled like in OSS in a systematic way. We are tracking this dependency partly via runtime-pm now, and I was wondering could we utilize runtime-pm to do this more cleanly (as sof-pci-dev owns the driver and initializes runtime pm and we have the appropriate runtime-pm calls in sof pcm.c already), but couldn't figure out a nice way.
So indeed, maybe @lgirdwood you have a recommendation? Otherwise I'd clean up the patch (clean the error handling) and send as RFC patch to alsa-devel to get comments.

Reply:
@kv2019i As SOF is still not upstreamed, maybe it is not proper to format a patch to discuss in ALSA mail list.
I will make a description about this and collect some suggestions in ALSA mail list. Could you give some advice there?

Reply:
@zhuyingjiang module refcount is the number of other modules that have a symbol dependency with this module. driver ref_count is the number of devices bound to this driver. 
ASoC modules should not be removable after machine driver binding unless the machine driver is removed first. So I guess this is about stopping removal of the machine driver whilst and PCM or kcontrol IO is active ? Can you confirm this is the bug you are fixing.  If so this needs fix at ASoC core level and not in SOF.